ext_proc: metadata and attributes

jbohanon opened the pull request:
Commit Message:
Introduce the ability to send dynamic metadata and attributes in the External Processing Request. Also implements the API for returning dynamic metadata as part of the External Processing Response
Additional Description:
Risk Level:
Low
Testing:
- Unit tests for processing request/response dynamic metadata
- Integration tests for request/response attributes

Docs Changes: TODO
Release Notes: N/A
Platform Specific Features: N/A
Fixes #19881

tyxia commented:
/assign @tyxia

rshriram commented on api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:
```diff
@@ -206,6 +204,17 @@ message ExternalProcessor {
+  //
+  // It works in a way similar to ``metadata_context_namespaces`` but allows envoy and external processing server to share the protobuf message definition
+  // in order to do a safe parsing.
+  repeated string typed_metadata_context_namespaces = 17;
```
Could we please add a knob for disabling this on Envoy side? just like disable immediate response etc..

Given the growing list of knobs, I wonder if this is an opportunity to consolidate all these enable/disable toggles into a message called serverCapabilities that has header mutation rules, allow_immediate_response, allow_dynamic_metadata, mode overrides, etc. And this capability proto should technically be advertised in the ProcessingRequest as well so that the ext_proc server knows what is and what is not allowed by the server.

Even if there is no appetite to do this capability thingy, we really need a boolean guard for every new piece of data that we will allow Envoy to accept from the Ext proc server. Some of our systems use this filter to talk to untrusted services and we would like to be strict about what things our data plane ingests from a remote ext_proc server. Hence the reason for toggles like immediate response, mode overrides, etc.

rshriram replied:
also, naming suggestion: accepted_typed_metadata_namespaces

jbohanon replied:
I've updated the proto changes to enable route-level overrides and in the process cleaned up this naming a bit.

rshriram commented on api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:
```diff
@@ -206,6 +204,17 @@ message ExternalProcessor {
+
+  // Specifies a list of metadata namespaces whose values, if present, will be passed to the
+  // ext_proc service as an opaque *protobuf::Struct*.
+  repeated string metadata_context_namespaces = 16;
```
I know you are reusing the field names from ext_authz but this is a chance to cleanup the poor naming.. how about dynamic_metadata_to_forward ?

stevenzzzz commented on source/extensions/filters/http/ext_proc/ext_proc.cc:
```diff
@@ -606,6 +677,85 @@ void Filter::onNewTimeout(const ProtobufWkt::Duration& override_message_timeout)
+  *req.mutable_metadata_context() = metadata_context;
+}
+
+void setDynamicMetadata(std::string ns, Http::StreamFilterCallbacks* cb,
```
This assumes we fully trust whatever is sent from the external server, can we add knobs to control (disable, filter ) what could be added into Envoy?

jbohanon replied:
I don't think this is necessary for a few reasons:

1. The external processing filter is listed as unknown security posture, only to be used with both trusted upstreams and downstreams
2. The external processing server has nearly arbitrary control over the request/response as it is, so I think it is safe to assume that it should be a trusted piece of code, at least as trusted as a custom Envoy filter which would be able to write dynamic metadata
3. The returned metadata is only able to be written into dynamic metadata in the ext_proc filter's own namespace.

rshriram replied:
1. its going to change soon. We have been vetting it and @yanjunxiang-google is going to flip this to trusted filter soon
2. Not true.. if you look at all the knobs we have added recently, we are really tying to limit what an external server can and cannot do so that those who want to use it against trusted systems can do so without knobs but those who need to use it against untrusted systems can rely on the knobs to tightly control what can/cannot be done.
3. follows 2 above in case of untrusted servers

jbohanon replied:
I've added the ability to disable writing the returned metadata. I think this is a good level of control for this initial implementation, but if more granularity is required for this work to merge, can you please point me to another similar implementation of the filtering?

yanjunxiang-google replied:
Thanks for adding the control knob. Yes, that's right.  We are going to flip this filter to be robust to untrusted downstream/upstream soon.

stevenzzzz reviewed (commented):
Thanks for working on it!

rshriram commented on api/envoy/service/ext_proc/v3/external_processor.proto:
```diff
@@ -158,7 +161,6 @@ message ProcessingResponse {
-  // [#not-implemented-hide:]
   // Optional metadata that will be emitted as dynamic metadata to be consumed by the next
   // filter. This metadata will be placed in the namespace ``envoy.filters.http.ext_proc``.
   google.protobuf.Struct dynamic_metadata = 8;
```
Could we please add a knob for disabling this on Envoy side? just like disable immediate response etc..

Given the growing list of knobs, I wonder if this is an opportunity to consolidate all these enable/disable toggles into a message called serverCapabilities that has header mutation rules, allow_immediate_response, allow_dynamic_metadata, mode overrides, etc. And this capability proto should technically be advertised in the ProcessingRequest as well so that the ext_proc server knows what is and what is not allowed by the server.

Even if there is no appetite to do this capability thingy, we really need a boolean guard for every new piece of data that we will allow Envoy to accept from the Ext proc server. Some of our products use this filter to talk to untrusted services and we would like to be strict about what things our data plane ingests from a remote ext_proc server. Hence the reason for toggles like immediate response, mode overrides, etc.

jbohanon replied:
I am happy to add a setting to allow disabling returned metadata to be written. I think that adding a capabilities message is outside the scope of this piece of work though

rshriram replied:
Given that this is already live in our systems, could you please turn this into a "enable_foo" knob rather than disable_foo ?

jbohanon replied:
Yes I have changed it to opt-in

jbohanon commented on source/extensions/filters/http/ext_proc/ext_proc.cc:
```diff
@@ -606,6 +677,85 @@ void Filter::onNewTimeout(const ProtobufWkt::Duration& override_message_timeout)
+      encoder_callbacks_->connection()->streamInfo().dynamicMetadata().filter_metadata();
+  const auto& request_encoder_metadata =
+      encoder_callbacks_->streamInfo().dynamicMetadata().filter_metadata();
+  for (const auto& context_key : config_->metadataContextNamespaces()) {
```
I have cut the lookups at least in half by dynamically determining from which callbacks we should be fetching the stream info to access the metadata. Currently, as stated in the comments, the request metadata will have precedent over the connection metadata.

RyanTheOptimist commented:
/assign @htuch

rshriram commented on api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:
```diff
@@ -206,6 +202,38 @@ message ExternalProcessor {
+
+  // If set to true, metadata returned from the external processing server will
+  // be written into the stream's dynamic metadata.
+  bool enable_returned_metadata = 2;
```
why cant this be namespaces_to_forward, namespaces_to_accept ?

jbohanon replied:
Is it desired to allow writing to arbitrary metadata namespaces?

jbohanon replied:
Is there a specific use-case you're thinking of? It seems to me that it is safer and more consistent to keep all of the returned metadata under the ext_proc namespace to prevent the external service from overwriting internal filters' data

nrjpoddar commented:
@rshriram long time!

Can we meet so we can move this forward and make sure we are not stepping on each other toes as it looks like both of our teams are working in this area.

phlax commented:
seems this has some design decisions to work through

in the meantime @jbohanon the CI fails look real

/wait

jbohanon commented:
Added additional testing to meet coverage and addressed the ASAN build. I'm not entirely sure what the Microsoft compiler is complaining about, but I added in a new `include` to hopefully resolve the issue

jbohanon commented:
Requesting a restart of [this check](https://dev.azure.com/cncf/envoy/_build/results?buildId=147690&view=logs&s=fbb7e3ed-ed2b-5ffd-2e3f-9e68630afbc3&j=8c169225-0ae8-53bd-947f-07cb81846cb5) as it appears to be an azure issue

jbohanon commented:
Building `arm64` image failed due to 503 from the ubuntu registry: https://dev.azure.com/cncf/envoy/_build/results?buildId=147690&view=logs&j=bd34f0c4-6c88-5907-5eba-b619c3aaa02d&t=1f15863c-20b8-505c-5551-f186b7e85332

This looks like another flake?

phlax commented:
> This looks like another flake?

yeah theres been a few today - some me breaking stuff, but in this case it was dockerhub going down, which has also been happening all day

phlax commented:
not much we can do docker is down https://statusgator.com/services/docker/docker-hub-registry

![image](https://github.com/envoyproxy/envoy/assets/454682/a18aacca-0d26-4231-a9e5-e902e28b4cf5)

jbohanon commented:
We don't really have a strong position on the implementation for the returned metadata, so I implemented writing to allowed namespaces to try to get this moving again. @rshriram if you get a chance to review the new changes I would really appreciate it. 

cc @htuch @stevenzzzz as well; this is ready for final review

tyxia reviewed (commented):
@jbohanon   Thanks for your contribution! This is definitely a valuable feature to have in ext_proc.

A high level question:  could you elaborate why CEL is introduced to this PR (for example, do you have an example of CEL expression)? Also from a quick scan,  I am curious if, for example, [StringMatcher](https://github.com/envoyproxy/envoy/blob/93b7427eb6bfcdd9c67b6723389b2f7cb05fa407/api/envoy/type/matcher/v3/string.proto#L21) meets your needs?

PS:  Regarding the way of introducing CEL, CEL matching is already available in Envoy as a shared lib/extension that can be used by individual filter( several filters are using it already).

Regardless, I want to know more about your motivation first. Thanks!

EItanya commented:
> A high level question: could you elaborate why CEL is introduced to this PR (for example, do you have an example of CEL expression)? Also from a quick scan, I am curious if, for example, [StringMatcher](https://github.com/envoyproxy/envoy/blob/93b7427eb6bfcdd9c67b6723389b2f7cb05fa407/api/envoy/type/matcher/v3/string.proto#L21) meets your needs?
> 
> PS: Regarding the way of introducing CEL, CEL matching is already available in Envoy as a shared lib/extension that can be used by individual filter( several filters are using it already).

For context, I wrote the attributes section of this PR, and @jbohanon is upstreaming it due to time constraints on my end. I chose CEL because it seemed to match the original intention of the API as found [in the proto](https://github.com/envoyproxy/envoy/blob/1a203945568db8ef28c3ebb6f3229d4f8d134fbe/api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto#L129C17-L142). That proto field points to the [attributes section of the docs](https://www.envoyproxy.io/docs/envoy/v1.27.0/intro/arch_overview/advanced/attributes.html#attributes), and when looking around to see how the attributes are typically implemented it seemed to be using CEL. CEL seems to be the only dynamic way of retrieving that data. In addition, CEL will allow more flexibility going forward to it seems like the correct implementation from that perspective as well.

jbohanon commented:
/retest [OOM flake](https://dev.azure.com/cncf/envoy/_build/results?buildId=148684&view=logs&jobId=b840a642-5ff3-5357-2e4b-e06e40b0cffd&j=b840a642-5ff3-5357-2e4b-e06e40b0cffd&t=67965174-5100-5631-9dc0-68b9f0aacb53)

jbohanon commented:
@tyxia @stevenzzzz @yanavlasov @rshriram Looking for final :eyes: on this PR please

jbohanon commented:
/retest

jbohanon commented:
> @tyxia @yanjunxiang-google can you please take a first pass on this?

@tyxia @yanjunxiang-google bumping this request.

rshriram commented:
The API changes LGTM. Thanks for accommodating our requests.

tyxia commented:
Apologize  for delay. I will take a look this week if that is OK.

I think introducing CEL is ok but i think there is a better way of introducing it. I will comment more once I get the chance to take a close look of full PR.

tyxia reviewed (commented):
I just looked at the matching part. I think we should introduce `xds.type.matcher.v3.Matcher` (i.e. Unified Matcher API) to ext_proc rather than CEL-only approach here, because:
1. It will be more powerful and flexible: With matching API, it  not only supports the CEL, but 
   also supports various matching API such as stringMatcher, etc. Also, it provides linear   
    and sublinear matching. Thus, the matching mechanism can be customized by ext_proc 
    end-user based on their use case.
2. It avoids expressive operations like CEL parsing(and checking if needed) in dataplane.  
    CEL team recommend doing this in control plane 
3. It will be a much cleaner implementation. I think the CEL related functionalities in this PR 
     are already available in the shared [cel_input](https://github.com/envoyproxy/envoy/tree/main/source/extensions/matching/http/cel_input) and [cel_match](https://github.com/envoyproxy/envoy/tree/main/source/extensions/matching/input_matchers/cel_matcher) extension. It can be reused in ext_proc by several lines of code.

You can refer to  [match_delegate](https://github.com/envoyproxy/envoy/blob/fc6342a95b48c3c8de0262085798c3a4369aa062/api/envoy/extensions/common/matching/v3/extension_matcher.proto#L34) and [rate_limit_quota](https://github.com/envoyproxy/envoy/blob/fc6342a95b48c3c8de0262085798c3a4369aa062/api/envoy/extensions/filters/http/rate_limit_quota/v3/rate_limit_quota.proto#L158) for examples. They are http filters using CEL already.

Furthermore, I think the matching-related stuff is technically separated from the main objective of this PR –- send/receive dynamic metadata and attributes. Therefore, I would recommend making this PR more self-contained and smaller by splitting the matching to a separate PR, if possible. This will also help robustness and review of this pr, which help move this PR forward.

jbohanon commented:
@rshriram I am on parental leave for a couple weeks more but I will ping my team to check on our priority for this

markdroth commented:
/lgtm api

jbohanon commented:
Per suggestion from @rshriram and others this PR will be closed in favor of smaller, self-contained PRs for the constituent features.

Dynamic Metadata: https://github.com/envoyproxy/envoy/pull/30747
Matching: TODO
Request/Response Attributes: TODO

jbohanon commented:
> I just looked at the matching part. I think we should introduce `xds.type.matcher.v3.Matcher` (i.e. Unified Matcher API) to ext_proc rather than CEL-only approach here, because:

I don't agree that this would be an appropriate place to introduce the Unified Matcher API. Primarily because we are not taking user-provided predicates and performing a disparate action based on the outcome, rather we are utilizing CEL to _look up_ the values represented by these attributes and send them to the processing server.
 
> 1. It will be more powerful and flexible: With matching API, it  not only supports the CEL, but
>    also supports various matching API such as stringMatcher, etc. Also, it provides linear
>    and sublinear matching. Thus, the matching mechanism can be customized by ext_proc
>    end-user based on their use case.

It is not relevant to support alternate methods of matching because [Attributes are implemented explicitly in CEL](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/common/expr/context.h) and not anywhere else

> 2. It avoids expressive operations like CEL parsing(and checking if needed) in dataplane.
>    CEL team recommend doing this in control plane

The expression parsing as implemented in this PR is done at config-time rather than at request-time. If absolutely necessary, we could look into parsing on the control plane and passing a parsed expression in via config, but this seems like it would add unnecessary readability issues into the config provided to Envoy. I admit the added complexity of the `USE_CEL_PARSER` ifdef is less than desirable but it is not without precedent in the project (see [access_loggers](https://github.com/envoyproxy/envoy/blob/main/source/extensions/access_loggers/filters/cel/config.cc) and [rate_limit_descriptors](https://github.com/envoyproxy/envoy/blob/main/source/extensions/rate_limit_descriptors/expr/config.cc))

> 3. It will be a much cleaner implementation. I think the CEL related functionalities in this PR
>    are already available in the shared [cel_input]
(https://github.com/envoyproxy/envoy/tree/main/source/extensions/matching/http/cel_input) and [cel_match](https://github.com/envoyproxy/envoy/tree/main/source/extensions/matching/input_matchers/cel_matcher) extension. It can be reused in ext_proc by several lines of code.

_IF_ we were to pass in parsed expressions, it is possible that we could use the cel_input's `HttpCelDataInput::get` method to create the activation, but that seems to be jumping through a lot of unnecessary hoops to get _just_ the activation, which still needs to be evaluated against the parsed expression(s). cel_match is not useful to us because again it is only outputting a boolean whereas we need the value of the attribute itself. 

> You can refer to [match_delegate](https://github.com/envoyproxy/envoy/blob/fc6342a95b48c3c8de0262085798c3a4369aa062/api/envoy/extensions/common/matching/v3/extension_matcher.proto#L34) and [rate_limit_quota](https://github.com/envoyproxy/envoy/blob/fc6342a95b48c3c8de0262085798c3a4369aa062/api/envoy/extensions/filters/http/rate_limit_quota/v3/rate_limit_quota.proto#L158) for examples. They are http filters using CEL already.

Each of these filters are using CEL to achieve a different result than we need here. Again, the goal here is to access the actual data which the attribute represents, not to assess it against a pre-determined predicate

tyxia commented:
@jbohanon  Sorry i have been busy with other projects and haven't got a chance to really look into this PR. But a few things from quick scan regarding your last comment
--->"The expression parsing as implemented in this PR is done at config-time rather than at request-time. "
I know it is config-time, but it is data plane config time which is still different from the control plane.

---> "this seems like it would add unnecessary readability issues into the config provided to Envoy"
There is no readability concern as you can always pass the config with simple original CEL string

Back to your motivation: The reason you are using CEL is only because  [Attributes are implemented explicitly in CEL](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/common/expr/context.h) and you use CEL for attributes look up?
